src/protocol/attribution/malicious.rs

```diff
@@ -41,24 +44,42 @@ where
     SemiHonestAdditiveShare<F>: Serializable,
 {
     let m_ctx = malicious_validator.context();
+    let m_binary_ctx = binary_malicious_validator.context();
 
     let futures = zip(
         repeat(
-            m_ctx
+            m_binary_ctx
                 .narrow(&Step::ComputeHelperBits)
-                .set_total_records(sorted_rows.len() - 1),
+                .set_total_records(sorted_match_keys.len() - 1),
         ),
-        sorted_rows.iter(),
+        sorted_match_keys.iter(),
     )
-    .zip(sorted_rows.iter().skip(1))
+    .zip(sorted_match_keys.iter().skip(1))
     .enumerate()
-    .map(|(i, ((m_ctx, row), next_row))| {
+    .map(|(i, ((c, row), next_row))| {
         let record_id = RecordId::from(i);
-        async move { bitwise_equal(m_ctx, record_id, &row.mk_shares, &next_row.mk_shares).await }
+        async move { bitwise_equal_gf2(c, record_id, &row, &next_row).await }
     });
+    let helper_bits_gf2 = try_join_all(futures).await?;
+
+    let validated_helper_bits_gf2 = binary_malicious_validator.validate(helper_bits_gf2).await?;
+
+    let hb_mod_conv_ctx = sh_ctx
+        .narrow(&Step::ModConvHelperBits)
+        .set_total_records(sorted_match_keys.len() - 1);
+    let semi_honest_helper_bits = try_join_all(validated_helper_bits_gf2.iter().enumerate().map(
+        |(i, gf2_bit)| {
+            let bit_triple = convert_bit_local::<F, Gf2>(sh_ctx.role(), 0, gf2_bit);
+            let record_id = RecordId::from(i);
+            let c = hb_mod_conv_ctx.clone();
+            async move { convert_bit(c, record_id, &bit_triple).await }
+        },
+    ))
+    .await?;
```

Comment:
This function is kinda long and awkward now.  Can you think of a way to label a chunk of this early stuff and put it in a function so that someone reading this function doesn't have to think about the details so much?

Calculating the helper bits is taking up a lot of space, so maybe just take all of that together.